### PR TITLE
Implement tuple interface for cuda vector types

### LINKED ISSUE #1406: [FEA]: CUDA vector types should implement the C++ tuple protocol
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

Not sure

### Is your feature request related to a problem? Please describe.

The CUDA types like `int3` defined in `<vector_types.h>` represent points. They are definitions in the global namespace and are automatically defined in every CUDA C++ program. There is no way for a CUDA C++ program to prevent their definition because these types participate in fundamental CUDA functionality like kernel launch and built-in variables. These are broadly-useful types with nice names and we are stuck with them in the global namespace.

The CUDA vector types are essentially tuples and they should behave like tuples. At the moment, they don't behave like tuples, because CUDA vector types don't implement the tuple protocol [1].

It's possible for third party libraries to implement the tuple protocol for these types themselves, but they must not do so. Firstly, because it is not their responsibility, but more importantly, because these implementations would inevitably collide when such libraries are combined.

CUDA has enhanced C++ support for these vector types in the past when constructors for these types were introduced. Making these types full-fledged tuples that can transparently participate in tuple-based libraries like CuTE and similar also makes sense.

[1] https://en.cppreference.com/w/cpp/utility/tuple/tuple-like

### Describe the solution you'd like

I believe that three things are necessary for CUDA vector types to implement the tuple protocol:

1. Provide customizations of `get`
2. Provide specializations of `std::tuple_element`
3. Provide specializations of `std::tuple_size`.

### Describe alternatives you've considered

It's possible for third party libraries to implement the solution described above, but they cannot coexist when those libraries are combined in a single program.

### Additional context

Many libraries, including the C++ Standard Library, present interfaces for _tuple-like_ types. It's inconvenient for users that the CUDA vector types do not conform to these interfaces.

### PR DESCRIPTION
**Pull request by miscco**
This specializes the `std::tuple_size` and `std::tuple_element` traits so that they are usable with cuda vector types.

We also provide overloads for `std::get`, which together enables structured bindings support in C++17 onwards.

Fixes #1406

### CODE REVIEW

**Comment by jaredhoberock** [comment]
Thanks for considering the feature request.

> We also provide overloads for std::get

To avoid putting additional overloads into `std`, it might be better to make `get` a member or `friend` function of these types.

**Comment by miscco** [comment]
> To avoid putting additional overloads of `std::get`, it might be better to make `get` a member or `friend` function of these types. There are lots of `std::get` functions, and introducing more may slow down overload resolution.

The preferred method would indeed be a hidden friend. Its a bit sad, that we need to define the `get` specializations twice due to the vector types being defined in the global namespace

File: libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_element.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+template <class VType, class BaseType, size_t Index>
+using expected_type = cuda::std::is_same<typename cuda::std::tuple_element<Index, VType>::type, BaseType>;
+
+template <class VType, class BaseType, size_t VSize, size_t Index, cuda::std::__enable_if_t<(Index < VSize), int> = 0>
+__host__ __device__ TEST_CONSTEXPR_CXX14 void test() {
+  assert((expected_type<VType, BaseType, Index>::value));
+  assert((expected_type<const VType, const BaseType, Index>::value));
+  assert((expected_type<volatile VType, volatile BaseType, Index>::value));
+  assert((expected_type<const volatile VType, const volatile BaseType, Index>::value));
```
Comment (griwes) [inline libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_element.pass.cpp:24]: Any reason for these not to be static asserts?

File: libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_size.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#include <cuda/std/tuple>
+
+#include "test_macros.h"
+
+template <class VType, size_t Size>
+__host__ __device__ TEST_CONSTEXPR_CXX14 void test() {
+  assert(cuda::std::tuple_size<VType>::value == Size);
+  assert(cuda::std::tuple_size<const VType>::value == Size);
+  assert(cuda::std::tuple_size<volatile VType>::value == Size);
+  assert(cuda::std::tuple_size<const volatile VType>::value == Size);
```
Comment (griwes) [inline libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_size.pass.cpp:20]: Ditto as in the tuple element test.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__tuple_dir/vector_types.h [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+      {};                                                                                              \
+      template <>                                                                                      \
+      struct tuple_size<const __name##__size> : _CUDA_VSTD::integral_constant<size_t, __size>          \
+      {};                                                                                              \
+      template <>                                                                                      \
+      struct tuple_size<volatile __name##__size> : _CUDA_VSTD::integral_constant<size_t, __size>       \
+      {};                                                                                              \
+      template <>                                                                                      \
+      struct tuple_size<const volatile __name##__size> : _CUDA_VSTD::integral_constant<size_t, __size> \
+      {};                                                                                              \
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__tuple_dir/vector_types.h:52]: The primary template for `tuple_size` should already be doing the stripping of cv qualifiers, why are we doing that
explicitly here? Ditto for `tuple_element`.

File: libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_element.pass.cpp (hunk 1 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_element.pass.cpp:24]: The only issue is that it is longer and I would need to add a `""` at the end

Comment (griwes) [reply libcudacxx/test/libcudacxx/cuda/tuple/vector_types_tuple_element.pass.cpp:24]: Let's make these into static asserts then.

**Review by wmaxey — APPROVED** [review]

**Pull request merged**